tools/WebcamReportTool/main.cpp

```diff
@@ -0,0 +1,189 @@
+#include <initguid.h>
+#define WIN32_LEAN_AND_MEAN
+
+#include <windows.h>
+#include <dshow.h>
+#include <cguid.h>
+
+#include <wil/com.h>
+#include <wil/resource.h>
+
+#include <iostream>
+#include <fstream>
+#include <string>
+
+#include "DirectShowUtils.h"
+
+std::ofstream& log()
+{
+    static std::ofstream report{ "webcam_diag_report.txt" };
+    return report;
+}
+
+const char* GetMediaSubTypeString(const GUID& guid)
+{
+    if (guid == MEDIASUBTYPE_RGB24)
+    {
+        return "MEDIASUBTYPE_RGB24";
+    }
+    else if (guid == MEDIASUBTYPE_RGB32)
+    {
+        return "MEDIASUBTYPE_RGB32";
+    }
+    else if (guid == MEDIASUBTYPE_YUY2)
+    {
+        return "MEDIASUBTYPE_YUY2";
+    }
+    else if (guid == MEDIASUBTYPE_MJPG)
+    {
+        return "MEDIASUBTYPE_MJPG";
+    }
+    else if (guid == MEDIASUBTYPE_NV12)
+    {
+        return "MEDIASUBTYPE_NV12";
+    }
+    else if (guid == MEDIASUBTYPE_NV11)
+    {
+        return "MEDIASUBTYPE_NV11";
+    }
+    else if (guid == MEDIASUBTYPE_YV12)
+    {
+        return "MEDIASUBTYPE_YV12";
+    }
+    else if (guid == MEDIASUBTYPE_YUYV)
+    {
+        return "MEDIASUBTYPE_YUYV";
+    }
+    else
+    {
+        return "MEDIASUBTYPE_UNKNOWN";
+    }
+}
+
+void LogMediaTypes(wil::com_ptr_nothrow<IPin>& pin)
+{
+    wil::com_ptr_nothrow<IEnumMediaTypes> mediaTypeEnum;
+    if (pin->EnumMediaTypes(&mediaTypeEnum); !mediaTypeEnum)
+    {
+        return;
+    }
+    ULONG _ = 0;
+    unique_media_type_ptr mt;
+    log() << "Supported formats:\n";
+    while (mediaTypeEnum->Next(1, wil::out_param(mt), &_) == S_OK)
+    {
+        if (mt->majortype != MEDIATYPE_Video)
+        {
+            continue;
+        }
+        auto format = reinterpret_cast<VIDEOINFOHEADER*>(mt->pbFormat);
+        if (!format->AvgTimePerFrame)
+        {
+            continue;
+        }
+        const auto formatAvgFPS = 10000000LL / format->AvgTimePerFrame;
+        log() << GetMediaSubTypeString(mt->subtype) << '\t' << format->bmiHeader.biWidth << "x"
+              << format->bmiHeader.biHeight << '\t' << formatAvgFPS << "fps\n";
```

Comment:
```suggestion
              << format->bmiHeader.biHeight << " -  " << formatAvgFPS << "fps\n";
```
In order to avoid this:
```
MEDIASUBTYPE_RGB24	960x720	30fps
MEDIASUBTYPE_RGB24	1024x576	30fps
MEDIASUBTYPE_RGB24	1184x656	30fps
MEDIASUBTYPE_RGB24	1280x720	30fps
MEDIASUBTYPE_RGB24	1280x960	30fps
MEDIASUBTYPE_UNKNOWN	640x480	30fps
```